xarray/conventions.py

```diff
@@ -878,4 +878,14 @@ def cf_encoder(variables, attributes):
                     if new_vars[bounds].attrs[attr] == var.attrs[attr]:
                         new_vars[bounds].attrs.pop(attr)
 
+    # remove _FillValue for coordinate variables as missing values are not permitted (issue #2037)
+    # a coordinate variable is a one-dimensional variable with the same name as its dimension
+    # see coordinate variable in http://cfconventions.org/cf-conventions/cf-conventions.html#terminology
+    for var in new_vars.keys():
+        if new_vars[var].dims == (var,) and new_vars[var].attrs.get("_FillValue"):
```

Comment:
1. Xarray does allow dimension coordinates with NaNs. So we'll need to check for the presence of nans using `np.any(duck_array_ops.isnull(new_vars[var]))`.  If there are no NaNs then we avoid the `_FillValue`.
2. Can we move this logic to https://github.com/pydata/xarray/blob/2c68a8b851581773733ab75dfd36590b528635e3/xarray/coding/variables.py#L144-L148
3. IIRC we always add `_FillValue` so the second half of the condition can be removed.

Reply:
From CF terminology:
> coordinate variable
We use this term precisely as it is defined in the [NUG section on coordinate variables](http://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_set_components.html#coordinate_variables). It is a one-dimensional variable with the same name as its dimension [e.g., time(time) ], and it is defined as a numeric data type with values that are ordered monotonically. **Missing values are not allowed in coordinate variables.**

I suppose the best is raise an Excepton `ValueError: Missing values are not allowed in coordinate variables` on the logic you indicated if `is a coordinate variable` and `there are null_values`.

Reply:
> 2. Can we move this logic to [xarray/xarray/coding/variables.py](https://github.com/pydata/xarray/blob/2c68a8b851581773733ab75dfd36590b528635e3/xarray/coding/variables.py#L144-L148)

I tired, but in the variable context, it is not possible to access to bound variables (which are kind of coordinate).
I also tried to modify here:
https://github.com/pydata/xarray/blob/33cdabd261b5725ac357c2823bd0f33684d3a954/xarray/conventions.py#L125-L133

But I think that overcomplicates the logic later if we have to search for the bounds.

Any ideas?